### ISSUE TITLE
Deadlock in win_iocp_io_context::shutdown() in case io_context is a thread_local variable

### ISSUE THREAD

**Issue by szenker** [chriskohlhoff/asio#520]
Hi,

the following code cause a deadlock (Win10) in method `win_thread::join()` which is invoked by `win_iocp_io_context::shutdown()`:

```cpp
#include <chrono>
#include <thread>
#include <iostream>
#include <boost/asio/io_context.hpp>
#include <boost/asio/steady_timer.hpp>

using namespace std::chrono_literals;

int main(int, char**)
{
    std::thread t([]()
    {
        thread_local boost::asio::io_context ctx;
        boost::asio::io_context::work dummy{ctx};
        boost::asio::steady_timer timer{ctx, 1s};
        timer.async_wait([ctx = &ctx](const auto&)
        {
            std::cout << "timer expired" << std::endl;
            ctx->stop();
        });
        ctx.run();
        std::cout << "finished" << std::endl;
    });
    t.join();
}
```

If you remove the `thread_local` specifier on line 13 then it works fine and terminates after 1 sec. In the documentation I didn't find any restriction, that `io_context` should not be allocated as `thread_local`.

Best Regards
Sebastian

**Comment by szenker** [comment]
Is it maybe related to #431 ?

**Comment by szenker** [comment]
I was using boost version 1.73.0

**Comment by youngwolf-project** [comment]
you should make a shared_ptr on boost::asio::io_context::work, and in another thread, when you want to end your application, call the shared_ptr's reset function (to release the internal boost::asio::io_context::work object) before the t.join() invocation.

BTW. boost::asio::io_context::work has been deprecated, use boost::asio::executor_work_guard instead.

**Comment by szenker** [comment]
@youngwolf-project My goal is not to terminate the second thread from the main (first) thread. Actually the 2nd thread is not needed at all, but in this case the example from above would work. It deems like it has something to do with which thread the `io_context` is instantiated on.

**Comment by youngwolf-project** [comment]
@szenker I meant the _dummy_ object prevented the return of _ctx.run()_, only after it destructed, the _ctx.run()_ can return and so does the _join_ invocation.
You can try make _dummy_ as a shared_ptr of boost::asio::io_context::work, and reset it in the timer's callback.

**Comment by szenker** [comment]
```cpp
#include <chrono>
#include <thread>
#include <iostream>
#include <boost/asio/executor_work_guard.hpp>
#include <boost/asio/io_context.hpp>
#include <boost/asio/steady_timer.hpp>

using namespace std::chrono_literals;

int main(int, char**)
{
    // boost::asio::detail::win_thread::set_terminate_threads(true);
    std::thread t([]()
    {
        thread_local boost::asio::io_context ctx;
        auto work = boost::asio::make_work_guard(ctx);
        boost::asio::steady_timer timer{ctx, 1s};
        timer.async_wait([&work](const auto&)
        {
            std::cout << "timer expired" << std::endl;
            work.reset();
        });
        ctx.run();
        std::cout << "finished" << std::endl;
    });
    t.join();
}
```

The output of this modified code is still the same (identical to the one above):
```
timer expired
finished
```

It still has the same problem, that it doesn't terminate and blocks forever inside `win_iocp_io_context::shutdown()`. The only work-around I have found so far, is to invoke `win_thread::set_terminate_threads(true)` but this isn't documented (public) API and I'm not sure about its side-effects. Does this maybe introduce memory leaks, etc.???

**Comment by youngwolf-project** [comment]
I used the standalone edition of asio with some small modifications on your codes (because I don't used c++14):

     // boost::asio::detail::win_thread::set_terminate_threads(true);
     std::thread tt([]()
     {
         thread_local asio::io_context ctx;
         auto work = asio::make_work_guard(ctx);
         asio::steady_timer timer{ctx, std::chrono::seconds(1)};
         timer.async_wait([&work](const asio::error_code&)
         {
             std::cout << "timer expired" << std::endl;
             work.reset();
         });
         ctx.run();
         std::cout << "finished" << std::endl;
     });
     tt.join();
     std::cout << "I see you." << std::endl;

It works properly:
timer expired
finished
I see you.

**Comment by szenker** [comment]
I tried running your example (above) using the latest standalone ASIO version (not boost). Same problem, it prints out:
```
timer expired
finished
```
but not
```
I see you.
```

Which version of Visual Studio did you use? I'm using:

* Visual Studio 16.5.4
* Microsoft C++ Compiler (cl) 19.25.28614
* Windows 10 (Version 1809 OS Build 17763.1282)
* current master branch of ASIO standalone

This bug occurs on Windows only. On Linux I cannot reproduce it because it blocks in `win_iocp_io_context::shutdown()`.

**Comment by youngwolf-project** [comment]
@szenker Oh, sorry, I used ubuntu 18.04, gcc 8.4.0
I'll try vc++ 2019 (with latest patches) because I have no other versions of vc++.

**Comment by youngwolf-project** [comment]
@szenker Yes, I reproduced the issue on windows, eliminate the _thread_local_ attribute can resolve the issue, I'm not sure it's an issue or just by design.

**Comment by szenker** [comment]
hmmm, I actually would like to have something like the following in my application:

```cpp
boost::asio::io_context& get_default_io_context()
{
  thread_local boost::asio::io_context ctx;
  return ctx;
}
```

That way, I effectively make the `io_context` a lazy initialized (per-thread) singleton and don't need to pass it to all places in my code which make use of boost asio TCP socket, websockets, etc. I didn't find any hint in the documentation, that this is not "allowed". A workaround is to invoke `boost::asio::detail::win_thread::set_terminate_threads(true);` but I'm not sure about the side-effects.
